### ISSUE TITLE
distributed_c10d.py Does Not Handle Custom Backends Correctly

### ISSUE THREAD

**Issue by Zha0q1** [pytorch/pytorch#92344]
### 🐛 Describe the bug

There is a on-going effort https://github.com/pytorch/pytorch/issues/86225 to decouple the ProcessGroup and Backend abstraction so that a single process group object can map to several backends based on the device type of the input and output tensors. 

`distributed_c10d.py` has been reworked as part of this effort. However, it seems like custom backends are not being handled correctly -- 
https://github.com/pytorch/pytorch/blob/88942a31991905ec1849f5b7fdc6266e7843d08d/torch/distributed/distributed_c10d.py#L1024
After the above line, we never set `backend_type` for custom backends and that lead to an invalid reference error here:
https://github.com/pytorch/pytorch/blob/88942a31991905ec1849f5b7fdc6266e7843d08d/torch/distributed/distributed_c10d.py#L1059-L1064.
Also, notice the check on line 1059. We should add in custom backend to the list, so that for a given process group we can map both "cpu" and "cuda" device types to the same custom backend object.

### Versions

master branch built from source
```
[pip3] numpy==1.23.5
[pip3] pytorch-triton==0.0.1
[pip3] torch==2.0.0a0+git692fd63
[pip3] torchaudio==2.0.0a0+4699ef2
[pip3] torchdata==0.6.0a0+a1612ee
[pip3] torchtext==0.15.0a0+f653dac
[pip3] torchvision==0.15.0a0+32d254b
[conda] magma-cuda117             2.6.1                         1    pytorch
[conda] mkl                       2022.2.1         h84fe81f_16997    conda-forge
[conda] mkl-include               2023.0.0         h84fe81f_25396    conda-forge
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-triton            0.0.1                    pypi_0    pypi
[conda] torch                     2.0.0a0+git692fd63          pypi_0    pypi
[conda] torchaudio                2.0.0a0+4699ef2          pypi_0    pypi
[conda] torchdata                 0.6.0a0+a1612ee          pypi_0    pypi
[conda] torchtext                 0.15.0a0+f653dac          pypi_0    pypi
[conda] torchvision               0.15.0a0+32d254b          pypi_0    pypi
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by H-Huang** [comment]
Hi @Zha0q1, thanks for raising the issue. Indeed custom backend support is still a work in progress. Backend is a new concept introduced as part of https://github.com/pytorch/pytorch/issues/86225, and the existing extensions extend the ProcessGroup class as shown here https://pytorch.org/tutorials/intermediate/process_group_cpp_extension_tutorial.html

We are planning to allow for C++ backend extension and provide a tutorial and documentation for this. Can you help me understand your use case and how you want to create a custom backend? Are you creating a custom backend in C++ or Python? What hardware are you using it for? Any information would be great, thanks!

**Comment by 0x6b64** [comment]
Thanks for the response @H-Huang . To answer your questions.
1)  how you want to create a custom backend?
We are part of AWS && we maintain the smddp library (https://docs.aws.amazon.com/sagemaker/latest/dg/data-parallel-modify-sdp-pt.html). We have already implemented the custom backend using the same strategy that you linked above.

2) Are you creating a custom backend in C++ or Python? What hardware are you using it for?
The custom backend (smddp) is in C++ && it is designed specifically to be performant on the AWS Infrastructure. (More details here: https://docs.aws.amazon.com/sagemaker/latest/dg/data-parallel-intro.html)

Our primary objective is to ensure that smddp continues to work seamlessly as a backend once PT 2.0 is launched. We have been prototyping with the nightly builds and this PR is one of the issues identified for successful integration.

A couple of questions for you:
1. smddp backend specifically relies on the GPU accelerator. However, (in ProcessGroup.hpp) existing GPU backend will only use a GPU tensor for the NCCL backend. Is there a reason for choosing this specific branching pattern. There is a similar technical issue in distributed_c10.py,  _get_pg_device method, which is not very flexible for custom backends who might also be using GPU's. Requesting you to provide your recommendation on the above pain point can be addressed.

3. The vision to allow one process group to use mulitple backend is clear from [the feature proposal thread](https://github.com/pytorch/pytorch/issues/86225). However, the implementation to map CPU and GPU tensors to different backend is still under development. [Reference](https://github.com/pytorch/pytorch/blob/ad782ff7df950dfda64e271dd4ee3c6128971103/torch/distributed/distributed_c10d.py#L246-L249). We would like to know the timeline of this development so that we can test our proposed use case, which is use SMDDP backend for GPU and Gloo for CPU, as soon as possible and revert back with any issue that requires co-design.

Happy to answer any questions/concern!

**Comment by Zha0q1** [comment]
Thanks @H-Huang for responding and @0x6b64 for providing details. One quick addition to what Karan shared - we have been providing a custom backend end at the c++ level by extending ProcessGroup with PyTorch 1.x. Now given the separation of Backend and ProcessGroup in PyTorch 2.x, we have switched to extending from the Backend base class.

**Comment by 0x6b64** [comment]
Hi @H-Huang , following up on the comments from before. :)  Also do you have a sense of when the refactor will be stable enough for us to start assessing our custom backend integration? As mentioned before, there are some places where strong assumptions are being made about exclusive use of GPU with the NCCL process group; and we want to be proactive about any code paths which need some refactor for custom process groups. Thank you so much!

**Comment by H-Huang** [comment]
Hi @Zha0q1 and @0x6b64, sorry for the delay. Thanks for the replies, I added responses inline:

> smddp backend specifically relies on the GPU accelerator. However, (in ProcessGroup.hpp) existing GPU backend will only use a GPU tensor for the NCCL backend. Is there a reason for choosing this specific branching pattern. There is a similar technical issue in distributed_c10.py, _get_pg_device method, which is not very flexible for custom backends who might also be using GPU's. Requesting you to provide your recommendation on the above pain point can be addressed.

We will continue to support allowing GPU tensors to be used with a GPU supporting backend (e.g. smddp). When users opt-in for multiple backends (backend argument not passed in), the default is CPU->gloo, GPU->nccl since PyTorch ships with both libraries, so users don't need to install anything extra.

I see the issue with `_get_pg_device` in your use case, its used in a few collectives which only checks if nccl is on the system. We can file a separate issue for this to track this so there isn't a NCCL dependency.

> We would like to know the timeline of this development so that we can test our proposed use case, which is use SMDDP backend for GPU and Gloo for CPU, as soon as possible and revert back with any issue that requires co-design.

We are very close to supporting this. I expect that it can be ready early next month.

> we have been providing a custom backend end at the c++ level by extending ProcessGroup with PyTorch 1.x. Now given the separation of Backend and ProcessGroup in PyTorch 2.x, we have switched to extending from the Backend base class. https://github.com/pytorch/pytorch/pull/92346 has the code changes in PyTorch that enable our current use case

Sounds good, we can get this PR merged first to unblock you. This may need to be updated if we want to allow for the use case mentioned above where you would like CPU->gloo, GPU->smddp, but we can merge it in for now.

> Also do you have a sense of when the refactor will be stable enough for us to start assessing our custom backend integration? As mentioned before, there are some places where strong assumptions are being made about exclusive use of GPU with the NCCL process group; and we want to be proactive about any code paths which need some refactor for custom process groups. Thank you so much!

GPU with NCCL is the most common use case, BUT we certainly don't intend to make any assumptions about the exclusive use of GPU with NCCL, are there any issues that you have seen other than `_get_pg_device`? If so then we can file issues to track this and correct this. We want to make sure ProcessGroup is able to incorporate all backends. The refactor is stable, but the dispatching to allow for custom backends is missing. I will keep this issue open and provide updates here, thanks for engaging and following along!
